#!/usr/bin/env Rscript

library(tidyverse)
library(fastbaps)

"Usage:
   run_fastbaps.R [--fasta <fasta> --out_dir <out_dir>]

Options:
   --fasta A genome alignment generated by snippy [default: snippy/gubbins.aln]
   --tree A phylogenetic tree [default: data/mashtree.dnd]
   --out_dir Output directory [default: data]
   
" -> doc

opts <- docopt::docopt(doc)

run_fastbaps <- function(fasta, tree, out_dir) {
  
  cols <- c("isolate", "level_1", "level_2")
  
  sparse_data <- fasta %>%
    fastbaps::import_fasta_sparse_nt(prior = "baps")
  
  rooted_tree <- tree %>%
    ggtree::read.tree() %>%
    phytools::midpoint.root()
  
  rooted_tree$tip.label <- rooted_tree$tip.label %>%
    str_replace(".contigs", "")
  
  keep <- base::intersect(rooted_tree$tip.label,
                          colnames(sparse_data$snp.matrix))
  
  rooted_tree <- ape::keep.tip(rooted_tree, keep)
  
  clusters <- fastbaps::best_baps_partition(sparse_data, rooted_tree) %>%
    as.data.frame() %>%
    rownames_to_column("isolate") %>%
    rename(cluster = 2)
  
  dir.create(out_dir, showWarnings = FALSE, recursive = TRUE)
  filename <- sprintf("%s/fastbaps.tsv", out_dir)
  write_tsv(clusters, filename)
  
}

if (sys.nframe() == 0) {
  run_fastbaps(opts$fasta, opts$out_dir)
}
