{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a7eadae6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.chdir('/Users/aaron/Desktop/MAC_paper/mac_evo/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af4054ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import pandas as pd\n",
    "\n",
    "ska_build = 'ska build -f ska/file_list_{x}.txt -k 31 -o ska/ska_index_{x} --threads 4'\n",
    "ska_align = 'ska align --min-freq 1 --filter no-filter ska/ska_index_{x}.skf -o ska/ska_alignment_{x}.aln --ambig-mask --threads 4'\n",
    "ska_map = 'ska map -o ska/ska_alignment_{x}.aln --ambig-mask --threads 4 databases/GCF_009741445.1/GCF_009741445.1_ASM974144v1_genomic.fna ska/ska_index_{x}.skf'\n",
    "run_gubbins = 'run_gubbins.py --prefix ska/gubbins_{x} ska/ska_alignment_{x}.aln --threads 12'\n",
    "snp_sites = f'docker run -v $PWD:/data staphb/snp-sites snp-sites -c {out_dir}/gubbins.filtered_polymorphic_sites.fasta > {out_dir}/gubbins.core.aln'\n",
    "snp_dists = f'docker run -v $PWD:/data staphb/snp-dists snp-dists {out_dir}/gubbins.core.aln > {out_dir}/snp_dists.txt'\n",
    "\n",
    "df = pd.read_csv('data/fastbaps.tsv', sep='\\t')\n",
    "clusters = [1, 2, 3, 5, 6, 7, 8, 11, 12, 13, 14, 15, 16, 17, 18]\n",
    "\n",
    "cmds = []\n",
    "for cluster in clusters:\n",
    "    name = 'fastbaps_cluster_' + str(cluster).zfill(2)\n",
    "    target = f'ska/gubbins_{name}.filtered_polymorphic_sites.fasta'\n",
    "    if not os.path.exists(target):\n",
    "        isolates = df[df['level_1'] == cluster]['isolate'].to_list()\n",
    "        output = file_list[file_list['isolate'].isin(isolates)]\n",
    "        output.to_csv(f'ska/file_list_{name}.txt', sep='\\t', index=False, header=False)\n",
    "        ## run the ska build command\n",
    "        build = ska_build.format(x=name)\n",
    "        map = ska_map.format(x=name)\n",
    "        gubbins = run_gubbins.format(x=name)\n",
    "        cmd = f'{build} && {map} && {gubbins}'\n",
    "        cmds.append(cmd)\n",
    "# write the commands to a file\n",
    "with open('ska_commands.sh', 'w') as f:\n",
    "    for cmd in cmds:\n",
    "        f.write(cmd + '\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ddda3565",
   "metadata": {},
   "outputs": [],
   "source": [
    "cids = [2, 3, 4, 5, 9, 10, 11, 18]\n",
    "with open('tidy_ska.sh', 'w') as f:\n",
    "    for cid in cids:\n",
    "        name = 'fastbaps_cluster_' + str(cid).zfill(2)\n",
    "        cmd = f'rm ska/*{name}*'\n",
    "        f.write(cmd + '\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "19589105",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_sites = 'docker run -v $PWD:/data staphb/snp-sites snp-sites -c ska/gubbins_{x}.filtered_polymorphic_sites.fasta > ska/core_{x}.aln'\n",
    "snp_dists = 'docker run -v $PWD:/data staphb/snp-dists snp-dists ska/core_{x}.aln > ska/core_{x}.dist'\n",
    "\n",
    "with open('run_snp_dists.sh', 'w') as f:\n",
    "    for cluster in clusters:\n",
    "        name = 'fastbaps_cluster_' + str(cluster).zfill(2)\n",
    "        target = f'ska/core_{name}.aln'\n",
    "        if not os.path.exists(target):\n",
    "            cmd_1 = snp_sites.format(x=name)\n",
    "            cmd_2 = snp_dists.format(x=name)\n",
    "            cmd = f'{cmd_1} && {cmd_2}'\n",
    "            f.write(cmd + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "47ed277c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import os\n",
    "\n",
    "os.chdir('/Users/aaron/Desktop/MAC_paper/mac_evo/')\n",
    "os.listdir('ska')\n",
    "\n",
    "fastas = glob.glob('ska/*.fasta')\n",
    "\n",
    "with open('file_list.txt', 'w') as f:\n",
    "    for fasta in fastas:\n",
    "        name = fasta.split('/')[-1].split('.')[0]\n",
    "        f.write(f'{name}\\t{fasta}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f25482d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## try ska with fastq files\n",
    "import glob\n",
    "import pandas as pd\n",
    "\n",
    "ska_build = 'ska build -f ska/file_list.txt -k 31 -o ska/index --threads 12'\n",
    "ska_align = 'ska align --min-freq 1 --filter no-filter ska_fq/ska_index_{x}.skf -o ska_fq/ska_alignment_{x}.aln --ambig-mask --threads 4'\n",
    "ska_map = 'ska map -o ska_fq/ska_alignment_{x}.aln --ambig-mask --threads 4 databases/GCF_009741445.1/GCF_009741445.1_ASM974144v1_genomic.fna ska_fq/ska_index_{x}.skf'\n",
    "run_gubbins = 'run_gubbins.py --prefix ska_fq/gubbins_{x} ska_fq/ska_alignment_{x}.aln --threads 12'\n",
    "\n",
    "df = pd.read_csv('data/metadata.tsv', sep='\\t')\n",
    "isolates = df['isolate'].to_list()\n",
    "\n",
    "if not os.path.exists('ska_fq'):\n",
    "    os.makedirs('ska_fq')\n",
    "\n",
    "with open('ska_fq/file_list.txt', 'w') as f:\n",
    "    for isolate in isolates:\n",
    "        reads = sorted(glob.glob('reads/processed/{}*'.format(isolate)))\n",
    "        output = f'{isolate}\\t{reads[0]}\\t{reads[1]}\\n'\n",
    "        f.write(output)\n",
    "\n",
    "df = pd.read_csv('data/fastbaps.tsv', sep='\\t')\n",
    "clusters = [1, 2, 3, 5, 6, 7, 8, 11, 12, 13, 14, 15, 16, 17, 18]\n",
    "\n",
    "file_list = pd.read_csv('ska_fq/file_list.txt', sep='\\t', header=None, names=['isolate', 'R1', 'R2'])\n",
    "\n",
    "cmds = []\n",
    "for cluster in clusters:\n",
    "    name = 'fastbaps_cluster_' + str(cluster).zfill(2)\n",
    "    target = f'ska_fq/gubbins_{name}.filtered_polymorphic_sites.fasta'\n",
    "    if not os.path.exists(target):\n",
    "        isolates = df[df['level_1'] == cluster]['isolate'].to_list()\n",
    "        output = file_list[file_list['isolate'].isin(isolates)]\n",
    "        output.to_csv(f'ska_fq/file_list_{name}.txt', sep='\\t', index=False, header=False)\n",
    "        ## run the ska build command\n",
    "        build = ska_build.format(x=name)\n",
    "        map = ska_map.format(x=name)\n",
    "        gubbins = run_gubbins.format(x=name)\n",
    "        cmd = f'{build} && {map} && {gubbins}'\n",
    "        cmds.append(cmd)\n",
    "# write the commands to a file\n",
    "with open('ska_fq_commands.sh', 'w') as f:\n",
    "    for cmd in cmds:\n",
    "        f.write(cmd + '\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92a71904",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('data/metadata.tsv', sep='\\t')\n",
    "isolates = df['isolate'].to_list()\n",
    "\n",
    "if not os.path.exists('ska_fq'):\n",
    "    os.makedirs('ska_fq')\n",
    "\n",
    "with open('ska_fq/file_list.txt', 'w') as f:\n",
    "    for isolate in isolates:\n",
    "        reads = sorted(glob.glob('reads/processed/{}*'.format(isolate)))\n",
    "        output = f'{isolate}\\t{reads[0]}\\t{reads[1]}\\n'\n",
    "        f.write(output)\n",
    "\n",
    "clusters = [1, 2, 3, 5, 6, 7, 8, 11, 12, 13, 14, 15, 16, 17, 18]\n",
    "\n",
    "cmds = []\n",
    "for cluster in clusters:\n",
    "    name = 'fastbaps_cluster_' + str(cluster).zfill(2)\n",
    "    target = f'ska_fq/gubbins_{name}.filtered_polymorphic_sites.fasta'\n",
    "    if not os.path.exists(target):\n",
    "        isolates = df[df['level_1'] == cluster]['isolate'].to_list()\n",
    "        output = file_list[file_list['isolate'].isin(isolates)]\n",
    "        output.to_csv(f'ska_fq/file_list_{name}.txt', sep='\\t', index=False, header=False)\n",
    "        ## run the ska build command\n",
    "        build = ska_build.format(x=name)\n",
    "        map = ska_map.format(x=name)\n",
    "        gubbins = run_gubbins.format(x=name)\n",
    "        cmd = f'{build} && {map} && {gubbins}'\n",
    "        cmds.append(cmd)\n",
    "# write the commands to a file\n",
    "with open('ska_fq_commands.sh', 'w') as f:\n",
    "    for cmd in cmds:\n",
    "        f.write(cmd + '\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6db957d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_sites = 'docker run -v $PWD:/data staphb/snp-sites snp-sites -c ska_fq/gubbins_{x}.filtered_polymorphic_sites.fasta > ska_fq/core_{x}.aln'\n",
    "snp_dists = 'docker run -v $PWD:/data staphb/snp-dists snp-dists ska_fq/core_{x}.aln > ska_fq/core_{x}.dist'\n",
    "\n",
    "with open('run_snp_dists.sh', 'w') as f:\n",
    "    for cluster in clusters:\n",
    "        name = 'fastbaps_cluster_' + str(cluster).zfill(2)\n",
    "        target = f'ska_fq/core_{name}.aln'\n",
    "        if not os.path.exists(target):\n",
    "            cmd_1 = snp_sites.format(x=name)\n",
    "            cmd_2 = snp_dists.format(x=name)\n",
    "            cmd = f'{cmd_1} && {cmd_2}'\n",
    "            f.write(cmd + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5af547a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "ska_build = 'ska build -f snippy_x_fastbaps/ska/file_list.txt -k 31 -o snippy_x_fastbaps/ska/core --threads 4'\n",
    "ska_align = 'ska align --min-freq 1 --filter no-filter snippy_x_fastbaps/ska/core.skf -o snippy_x_fastbaps/ska/core.aln --ambig-mask --threads 4'\n",
    "\n",
    "run_gubbins = 'run_gubbins.py --prefix ska_fq/gubbins_{x} ska_fq/ska_alignment_{x}.aln --threads 12'\n",
    "\n",
    "\n",
    "os.chdir('/Users/aaron/Desktop/MAC_paper/mac_evo/')\n",
    "\n",
    "if not os.path.exists('snippy_x_fastbaps/ska'):\n",
    "    os.makedirs('snippy_x_fastbaps/ska')\n",
    "\n",
    "fastas = [x.as_posix() for x in Path('snippy_x_fastbaps').rglob('*.fasta')]\n",
    "for fasta in fastas:\n",
    "    for record in SeqIO.parse(fasta, \"fasta\"):\n",
    "        isolate = record.id\n",
    "        if not 'Reference' in isolate:\n",
    "            sequence = str(record.seq)\n",
    "            with open(f'snippy_x_fastbaps/ska/{isolate}.fasta', 'w') as f:\n",
    "                f.write(f'>{isolate}\\n{sequence}\\n')\n",
    "\n",
    "files = [x.as_posix() for x in Path('snippy_x_fastbaps/ska').glob('*.fasta')]\n",
    "with open('snippy_x_fastbaps/ska/file_list.txt', 'w') as f:\n",
    "    for file in files[0:10]:\n",
    "        isolate = file.split('/')[-1].split('.')[0]\n",
    "        f.write(f'{isolate}\\t{file}\\n')\n",
    "\n",
    "with open('run_ska.sh', 'w') as f:\n",
    "    cmd = f'{ska_build} && {ska_align}\\n'\n",
    "    f.write(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17bdea4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## try ska with fastq files\n",
    "import glob\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "os.chdir('/Users/aaron/Desktop/MAC_paper/mac_evo/')\n",
    "\n",
    "if not os.path.exists('ska'):\n",
    "    os.makedirs('ska')\n",
    "\n",
    "ska_build = 'ska build -f ska/file_list.txt -k 31 -o ska/index --threads 12 -v'\n",
    "ska_align = 'ska align --min-freq 1 --filter no-filter ska/index.skf -o ska/alignment.aln --ambig-mask --threads 12 -v'\n",
    "fastbaps = 'Rscript src/run_fastbaps.R && Rscript src/list_fastbaps_clusters.R'\n",
    "\n",
    "with open('ska/file_list.txt', 'w') as f:\n",
    "    assemblies = [x for x in glob.glob('assemblies/*') if not 'Reference' in x]\n",
    "    for assembly in assemblies:\n",
    "        isolate = os.path.basename(assembly.split('.')[0])\n",
    "        reads = '\\t'.join(sorted(glob.glob(f'reads/processed/{isolate}*')))\n",
    "        row = f'{isolate}\\t{reads}\\n'\n",
    "        f.write(row)\n",
    "\n",
    "with open('run_ska.sh', 'w') as f:\n",
    "    f.write(ska_build + '\\n')\n",
    "    f.write(ska_align + '\\n')\n",
    "    f.write(fastbaps + '\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a0a37469",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "ska_build = 'ska build -f {out_dir}/file_list.txt -k 31 -o {out_dir}/index --threads {threads} -v'\n",
    "ska_map = 'ska map -o {out_dir}/alignment.aln --ambig-mask --threads {threads} {ref} {out_dir}/index.skf'\n",
    "run_gubbins = 'run_gubbins.py --prefix {out_dir}/gubbins {out_dir}/alignment.aln --threads {threads}'\n",
    "snp_sites = 'docker run -v $PWD:/data staphb/snp-sites snp-sites -c {out_dir}/gubbins.filtered_polymorphic_sites.fasta > {out_dir}/gubbins.core.aln'\n",
    "snp_dists = 'docker run -v $PWD:/data staphb/snp-dists snp-dists {out_dir}/gubbins.core.aln > {out_dir}/snp_dists.txt'\n",
    "\n",
    "def run_ska_x_fastbaps(out_dir='ska_x_fastbaps', threads=os.cpu_count()):\n",
    "    if not os.path.exists(out_dir):\n",
    "        os.makedirs(out_dir)\n",
    "    dat = pd.read_csv('data/fastbaps.tsv', sep='\\t')\n",
    "    clusters = [int(x) for x in open('data/fastbaps_clusters.txt', 'r').read().splitlines()]\n",
    "    cmds = []\n",
    "    for cluster in clusters:\n",
    "        name = 'fastbaps_cluster_' + str(cluster).zfill(2)\n",
    "        outdir = f'{out_dir}/{name}'\n",
    "        if not os.path.exists(outdir):\n",
    "            os.makedirs(outdir)\n",
    "        isolates = sorted(dat[dat['level_1'] == int(cluster)]['isolate'].to_list())\n",
    "        targets = [f'{outdir}/index.skf',\n",
    "                   f'{outdir}/alignment.aln',\n",
    "                   f'{outdir}/gubbins.filtered_polymorphic_sites.fasta',\n",
    "                   f'{outdir}/gubbins.core.aln',\n",
    "                   f'{outdir}/snp_dists.txt']\n",
    "        if not os.path.exists(targets[0]):\n",
    "            with open(f'{outdir}/file_list.txt', 'w') as file_list:\n",
    "                for isolate in isolates:\n",
    "                    reads = '\\t'.join(sorted(glob.glob(f'reads/processed/{isolate}*')))\n",
    "                    row = f'{isolate}\\t{reads}\\n'\n",
    "                    file_list.write(row)\n",
    "            skabuild = ska_build.format(out_dir=outdir, threads=threads)\n",
    "            cmds.append(skabuild)\n",
    "        if not os.path.exists(targets[1]):\n",
    "            ref = 'assemblies/{}.contigs.fasta'.format(isolates[0])\n",
    "            skamap = ska_map.format(out_dir=outdir, threads=threads, ref=ref)\n",
    "            cmds.append(skamap)\n",
    "        if not os.path.exists(targets[2]):\n",
    "            rungubbins = run_gubbins.format(out_dir=outdir, threads=threads)\n",
    "            cmds.append(rungubbins)\n",
    "        if not os.path.exists(targets[3]):\n",
    "            snpsites = snp_sites.format(out_dir=outdir)\n",
    "            cmds.append(snpsites)\n",
    "        if not os.path.exists(targets[4]):\n",
    "            snpdists = snp_dists.format(out_dir=outdir)\n",
    "            cmds.append(snpdists)\n",
    "    with open('run_ska_x_fastbaps.sh', 'w') as f:\n",
    "        [f.write(cmd + '\\n') for cmd in cmds]\n",
    "\n",
    "run_ska_x_fastbaps()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "34ed2240",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.chdir('/Users/aaron/Desktop/MAC_paper/mac_evo/')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
