#!/usr/bin/env Rscript

library(tidyverse)

"Usage:
   run_fastbaps.R [--fasta <fasta> --out_dir <out_dir>]

Options:
   --fasta A genome alignment generated by snippy [default: ska/alignment.aln]
   --out_dir Output directory [default: data]
   
" -> doc

opts <- docopt::docopt(doc)

run_fastbaps <- function(fasta, out_dir) {
  
  cols <- c("isolate", "level_1", "level_2")
  
  clusters <- fasta %>%
    fastbaps::import_fasta_sparse_nt() %>%
    fastbaps::optimise_prior(type = "optimise.symmetric") %>%
    fastbaps::multi_res_baps() %>%
    setNames(cols)
  
  dir.create(out_dir, showWarnings = FALSE, recursive = TRUE)
  filename <- sprintf("%s/fastbaps.tsv", out_dir)
  write_tsv(clusters, filename)
  
}

if (sys.nframe() == 0) {
  run_fastbaps(opts$fasta, opts$out_dir)
}
